File: 알고리즘 문제풀이/프로그래머스/구명보트.py
"""
[70,50,80,50]  limit:100
50 50 70 80
50 50
70
80

[50,70,80]

"""
def solution(people, limit):
    answer,le,ri = 0,0,len(people)-1
    people=sorted(people)
    while le<ri:
        if people[le]+people[ri]>limit:
            answer+=1
        elif people[le]+people[ri]<=limit:
            le+=1
            answer+=1
        ri-=1
    if le==ri:
        answer+=1
    return answer

def solution2(people,limit):
    answer=0
    people=sorted(people)
    while people:
        if len(people)==1:
            answer+=1
            break
        #제일 작은 값과 가장 큰값의 합이 limit를 넘는지 안넘는지 검사 진행
        if people[0]+people[-1]>limit: 
            people.pop()
        elif people[0]+people[-1]<=limit:
            people.pop()
            people.pop(0)
        answer+=1
    return answer

File: 알고리즘 문제풀이/프로그래머스/구명보트.js
//한 번에 최대 2명씩 밖에 탈 수 없고, 무게 제한도 있습니다.
function solution(people, limit) {
  let answer = 0,
    le = 0,
    ri = people.length - 1;
  people.sort((x, y) => (x > y ? 1 : x === y ? 0 : -1));
  while (le < ri) {
    if (people[le] + people[ri] <= limit) {
      le++;
      ri--;
    } else {
      ri--;
    }
    answer += 1;
  }
  return le === ri ? answer + 1 : answer;
}
